src/centralized_telescope/proof.rs

```diff
@@ -248,21 +275,25 @@ impl Proof {
     /// prehashing S_p
     fn bin_hash(set_size: u64, retry_counter: u64, element: Element) -> Option<u64> {
         let retry_bytes: [u8; 8] = retry_counter.to_be_bytes();
-        let mut hasher = Blake2s256::new();
-        hasher.update(b"Telescope-bin_hash");
-        hasher.update(retry_bytes);
-        hasher.update(element);
-        let digest: Hash = hasher.finalize().into();
-        sample::sample_uniform(&digest, set_size)
+        let digest = H::new()
+            .chain_update(b"Telescope-bin_hash")
+            .chain_update(retry_bytes)
+            .chain_update(element)
+            .finalize()
+            .to_vec();
+        let hash = truncate(&digest);
+        sample::sample_uniform(&hash, set_size)
     }
 
     /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0
     /// otherwise
-    fn proof_hash(valid_proof_probability: f64, r: &Round) -> bool {
-        let mut hasher = Blake2s256::new();
-        hasher.update(b"Telescope-proof_hash");
-        hasher.update(r.hash);
-        let digest: Hash = hasher.finalize().into();
-        sample::sample_bernoulli(&digest, valid_proof_probability)
+    fn proof_hash(valid_proof_probability: f64, r: &Round<H>) -> bool {
+        let digest = H::new()
+            .chain_update(b"Telescope-proof_hash")
+            .chain_update(r.hash)
+            .finalize()
+            .to_vec();
+        let hash = truncate(&digest);
```

Comment:
Looks like that old method doesn't work since `GenericArray` only implements `From` for certain array sizes. https://docs.rs/generic-array/0.14.7/src/generic_array/impls.rs.html#183-216

At the same time truncation is sketchy, the hash function could have a very small output in principle. I think a better solution, though not ideal, would be to replace `.to_vec()` with `.to_slice()`, then runtime assert that the slice size is at least 16 bytes and pass an array of 16 bytes to sample.

Reply:
Good call!